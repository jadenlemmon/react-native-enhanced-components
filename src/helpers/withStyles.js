import React from 'react';
import { pick, get } from 'lodash';
import gs from '../styles';

/**
 * Map props to global styles. Use value uses value of prop
 * @param {*} props
 * @param {*} useValue
 */
const mapStyle = (props, useValue, fromStyles) => Object.keys(props).map(key => (useValue ? fromStyles[`${key}${props[key]}`] : fromStyles[`${key}`]));

export default function withStyles(
  WrappedComponent,
  styles,
  useValues = false,
  fromStyles = gs,
) {
  return class extends React.Component {
    getStyles() {
      // get other styles
      const otherStyles = get(this.props, 'style', {});

      const grabbedStyles = mapStyle(pick(this.props, styles), useValues, fromStyles);

      // grab other styles
      const other = Array.isArray(otherStyles) ? otherStyles : [otherStyles];

      // merge all
      return [...grabbedStyles, ...other];
    }

    render() {
      return <WrappedComponent {...this.props} style={this.getStyles()} />;
    }
  };
}
