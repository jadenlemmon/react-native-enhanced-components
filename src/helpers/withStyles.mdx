---
name: With Styles HOC
---

import { Playground, PropsTable } from 'docz';
import { TouchableOpacity, Text, View } from 'react-native';
import withStyles from './withStyles';
import withSpacing from './withSpacing';
import Block from '../components/Block';

# With Styles HOC

## Minimel Example

Creating a button that exposes as many style props as you would like that can be reused with ease throughout your app.

<Playground>
  {() => {
    const buttonStyles = {
      tomato: {
        backgroundColor: 'tomato',
      },
      purple: {
        backgroundColor: 'purple',
      },
      grey: {
        backgroundColor: 'grey',
      },
      rounded: {
        borderRadius: 10,
      },
    };
    const MyCustomButton = withSpacing(withStyles(View, buttonStyles));
    return (
      <Block flex row justifyBetween>
        <MyCustomButton tomato rounded p={10} style={{ width: 200 }}>
          <Text style={{ color: 'white', textAlign: 'center' }}>My Button</Text>
        </MyCustomButton>
        <MyCustomButton purple rounded p={10} style={{ width: 200 }}>
          <Text style={{ color: 'white', textAlign: 'center' }}>My Button</Text>
        </MyCustomButton>
        <MyCustomButton grey p={10} style={{ width: 200 }}>
          <Text style={{ color: 'white', textAlign: 'center' }}>My Button</Text>
        </MyCustomButton>
      </Block>
    );
}}

</Playground>
