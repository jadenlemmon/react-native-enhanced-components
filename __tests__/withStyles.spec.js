import React from 'react';
import renderer from 'react-test-renderer';
import { View, StyleSheet } from 'react-native';
import withStyles from '../src/helpers/withStyles';

const styles = StyleSheet.create({
  foo: {
    backgroundColor: 'tomato',
  },
});

const FooBar = withStyles(View, styles);

test('it should render component with backgroundColor', () => {
  const tree = renderer.create(<FooBar foo />).toJSON();
  expect(tree.props.style[0]).toHaveProperty('backgroundColor');
  expect(tree).toMatchSnapshot();
});
