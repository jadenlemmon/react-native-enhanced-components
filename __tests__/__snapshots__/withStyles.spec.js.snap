// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render component with backgroundColor 1`] = `
<View
  foo={true}
  style={
    Array [
      Object {
        "backgroundColor": "tomato",
      },
      Object {},
    ]
  }
/>
`;
